python/cuml/dask/decomposition/pca.py

```diff
@@ -0,0 +1,472 @@
+
+        comms = CommsContext(comms_p2p=False)
+        comms.init(workers=workers)
+
+        worker_info = comms.worker_info(comms.worker_addresses)
+
+        key = uuid1()
+        partsToRanks = [(worker_info[wf[0]]["r"], self.client.submit(
+            PCA._func_get_size,
+            wf[1],
+            workers=[wf[0]],
+            key="%s-%s" % (key, idx)).result())
+            for idx, wf in enumerate(gpu_futures)]
+
+        N = X.shape[1]
+        M = reduce(lambda a,b: a+b, map(lambda x: x[1], partsToRanks))
+
+        key = uuid1()
+        self.pca_models = [(wf[0], self.client.submit(
+            PCA._func_create_model,
+            comms.sessionId,
+            wf[1],
+            **self.kwargs,
+            workers=[wf[0]],
+            key="%s-%s" % (key, idx)))
+            for idx, wf in enumerate(worker_to_parts.items())]
+
+        key = uuid1()
+        pca_fit = dict([(worker_info[wf[0]]["r"], self.client.submit(
+            PCA._func_fit,
+            wf[1],
+            M, N,
+            partsToRanks,
+            worker_info[wf[0]]["r"],
+            _transform,
+            key="%s-%s" % (key, idx),
+            workers=[wf[0]]))
+            for idx, wf in enumerate(self.pca_models)])
+
+        wait(list(pca_fit.values()))
+        raise_exception_from_futures(list(pca_fit.values()))
+
+        comms.destroy()
+
+        self.local_model = self.client.submit(PCA._func_get_first,
+                                              self.pca_models[0][1]).result()
+
+        self.components_ = self.local_model.components_
+        self.explained_variance_ = self.local_model.explained_variance_
+        self.explained_variance_ratio_ = self.local_model.explained_variance_ratio_
+        self.singular_values_ = self.local_model.singular_values_
+        self.noise_variance = self.local_model.noise_variance_
+
+        out_futures = []
+        if _transform:
+            completed_part_map = {}
+            for rank, size in partsToRanks:
+                if rank not in completed_part_map:
+                    completed_part_map[rank] = 0
+           
+                f = pca_fit[rank]
+                out_futures.append(self.client.submit(
+                    PCA._func_get_idx, f, completed_part_map[rank]))
+
+                completed_part_map[rank] += 1
+
+            return to_dask_cudf(out_futures)
+
+    def _transform(self, X):
+        gpu_futures = self.client.sync(extract_ddf_partitions, X, agg=False)
+
+        worker_to_parts = OrderedDict()
+        for w, p in gpu_futures:
+            if w not in worker_to_parts:
+                worker_to_parts[w] = []
+            worker_to_parts[w].append(p)
+
+        workers = list(map(lambda x: x[0], gpu_futures))
+
+        comms = CommsContext(comms_p2p=False)
+        comms.init(workers=workers)
+
+        worker_info = comms.worker_info(comms.worker_addresses)
+
+        key = uuid1()
+        partsToRanks = [(worker_info[wf[0]]["r"], self.client.submit(
```

Comment:
This all seems completely unnecessary and causes an extra NCCL initialization, which takes several seconds on 2 DGX1s. You are already moving the trained model on the client on line 282. Why not just broadcast that out to do the transform embarrassingly parallel and without the added complexity / synchronization of creating a new communicator?

K-Means has a great example of this here: https://github.com/rapidsai/cuml/blob/branch-0.11/python/cuml/dask/cluster/kmeans.py#L237

Reply:
NCCL init code will be removed. I don't actually do any communication in C++ during transform operation.

This transform implementation in K-Means doesn't seem to me an efficient way to do it unless I miss something here. 

Is it calling the transform function for every partition? If yes, all the transform operation per GPU will be serialized due to the same stream.

I guess you are concatenating the different partitions here which is an extra copying operation. Something we don't want to do. 

I keep that redundant C++ implementation of transform operation to prevent these kind of copying operations.

Reply:
You bring up a great point, actually. We can get even more parallelism by running the transform in separate streams. 

I’ve filed #1331 to handle this case, which should also enable us to do it right in the Python layer, since most of our ML algorithms are going to need this.

Reply:
Sounds good. Thanks.

Reply:
NCCL code is removed from cython. FYI.